### PR TITLE
[google_maps_flutter] Add marker clustering support - android implementation

### PR DESCRIPTION
**Pull request by jokerttu**
This PR introduces support for marker clustering for Android platform

An example usage is available in the example application at `./packages/google_maps_flutter/google_maps_flutter_android/example` on the page `Manage clustering`

This is prequel PR for: https://github.com/flutter/packages/pull/4319
and sequel PR for: https://github.com/flutter/packages/pull/6158

Containing only changes to `google_maps_flutter_android` package.

Follow up PR will hold the app-facing plugin implementation.

Linked issue: https://github.com/flutter/flutter/issues/26863

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by reidbaker** [comment]
Did not mean to close.

**Review by reidbaker — COMMENTED** [review]
Partial review. Sorry you waited so long this slipped though my radar. Will continue the review tomorrow.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/MarkersControllerTest.java
Comment (camsim99) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/MarkersControllerTest.java:1]: Can you test the functionality you added to `onMapsMarkerTap`, `onClusterItemMarker`, `changeMarker` here or wherever is appropriate? Seems like they are called by `GoogleMapsController`, but testing there would require faking the method calls I assume.

**Review by camsim99 — COMMENTED** [review]
Looks great! Just left some comments about nits, tests, and naming :) Would not merge until @reidbaker gives a second review though

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:82]: How would you like me to change this?
Exceptions thrown in the native code are converted to PlatformExceptions by the method channel and this is handled as an PlatformException at dart code.

Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:82]: IllegalArgumentExceptions are used in the Android platform on similar places.
Exceptions thrown in the method channel method handlers are converted to PlatformExceptions by the method channel code:

E/flutter (12022): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: **PlatformException**(error, clusterManagerId was null, null, java.lang.IllegalArgumentException: clusterManagerId was null

As clusterManagerUpdates should be created by the [ClusterManagerUpdates](https://github.com/flutter/packages/blob/dce6f0c91fcf01ee3a67e3475155ec9f16ce2a66/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager_updates.dart#L11) class this error should not be never seen by the developers.

How would you like me to change this handling?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:148]: Current implementation of the google maps plugin uses custom serialisation for method channel communication.
These strings are hardcoded through the native implementation matching the data structure serialised at the dart layer.

Corresponding mapping for this is done here:
https://github.com/flutter/packages/blob/dce6f0c91fcf01ee3a67e3475155ec9f16ce2a66/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/cluster_manager.dart#L62

Should I add this link to before the line, as this might get outdated (even with the ref) at some point.
The current communication layer relies a lot of these hardcoded mapping keys.
Moving the plugin method channel over pigeon in the future would improve the type-safety a lot.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:166]: There is no reason to move the `setGoogleMapListener(this)` listener initialization to different position and was only done to re-organize listeners to same place and improve readability; I reverted this change as it is not needed by this PR.

Some of the listeners are removed (eg. `googleMap.setOnInfoWindowClickListener(this)`) as those are handled by the markerCollection listeners set in `setMarkerCollectionListener`

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:220]: renamed the method and interface to match the method call where it is called: `onClusterItemRendered`

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterItemListener.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterItemListener.java:1]: This method combines both:
ClusterManagersController.OnClusterItemMarker<MarkerBuilder>,
and 
ClusterManager.OnClusterItemClickListener<MarkerBuilder>
under same listener.

Listener with type ClusterItemListener is registered for clusterManagerController.
It is possible to have two listenerer objects instead of this combinations, and two setter functions to set the listener.

Instead of method   
```
void setClusterItemListener(@Nullable ClusterItemListener listener) {
    clusterItemListener = listener;
    initListenersForClusterManagers(this, listener);
}
```

there need to be two methods:

```
  void setOnClusterItemClickListener(@Nullable ClusterManager.OnClusterItemClickListener<MarkerBuilder> listener) {
    onClusterItemClickListener= listener;
    initListenersForClusterManagers(this, listener);
  }
```
  
  and
  
```
  void setOnClusterItemRenderedListener(@Nullable ClusterManagersController.OnClusterItemMarker<MarkerBuilder>listener) {
    onClusterItemRenderedListener = listener;
  }
```

Is this approach better?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterListener.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterListener.java:1]: Removed this file

File: packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/example/integration_test/google_maps_tests.dart:1]: I don't know good way to do this, as this method is called by the the google maps sdk when the marker on map is clicked.
It could be possible to position marker in middle of the map and try to tap that location programmically, but that sounds a bit flaky?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java:78]: When marker does not have clusterManagerId it is added to the map by MarkersController and need to be removed from the map. If marker is set to be controller by the clusterManager, the marker is removed from clusterManager istead.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:144]: Added comment

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerBuilder.java
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerBuilder.java:40]: Why is positions updated on line 30 and here?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerBuilder.java
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerBuilder.java:29]: So that I can understand can you explain how markerOptionsToUpdate updates information?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerController.java
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerController.java:24]: Can you add a comment here explaining why a weakreference is needed? I believe that you would not have added the complexity if it was not needed but it helps when someone else comes to modify the code.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java:104]: Consider including the markerId in this error message.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/ConvertTest.java
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/ConvertTest.java:38]: This test could be easier to read if you had position1/2/3 defined a the top then passed those positions into the constructor and compared to those positions in the asserts.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java:196]: Specifically here but a comment more broadly about the java side code and tests. Please add some more coverage of the non happy path cases in tests.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:187]: Document/test the behavior here if onClusterTap is called with an invalid mapId

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/utils/cluster_manager_utils.dart
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/utils/cluster_manager_utils.dart:8]: Please add cluster_manager_utils_test that covers these methods.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:163]: (Tarrin just recently landed support for channel suffixes in Pigeon, which unblocks converting maps, so I'm hoping to do the conversion this quarter.)

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerBuilder.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerBuilder.java:29]: Google maps markers are created with MarkerOptions object.
Items added to clusterManager are instances of MarkerBuilder that extends the ClusterItem class. The clustering process uses data from this MarkerBuilder instance to update the MarkerOptions, that it then uses to create or re-create the marker to the map.

It is important to store this data to MarkerBuilder object instance per each marker to allow to update marker data even when marker instance is not drawn on the map.
Here is clustering callback that uses this update method tuo pass the data from the MarkerBuilder to the markerOptions instance, the clustermanager uses to create the visible marker on the map.

```java
    @Override
    protected void onBeforeClusterItemRendered(
        @NonNull MarkerBuilder item, @NonNull MarkerOptions markerOptions) {
      // Builds new markerOptions for new marker created by the ClusterRenderer under
      // ClusterManager.
      item.update(markerOptions);
    }
```

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerController.java:24]: Isn't the comment on few lines above is enough explaining the need for the weakReference?
As the clustering process might remove the marker from the map (when clustered), the marker should not be accessed anymore in MarkerController. Instead markerbuilder is updated, so that if the marker is declustered, it is used to update the marker settings before  drawing.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java]: Reverted the change, as just noticed that this changes the existing behaviour and is breaking change.
Eventhough it would be nice to give extra information for the developers that the error could be caused because marker is removed by the clusterManager, changing the existing API might not be an option here.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java
Comment (jokerttu) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java]: Reverted the change, as just noticed that this changes the existing behaviour and is breaking change.
Eventhough it would be nice to give extra information for the developers that the error could be caused because marker is removed by the clusterManager, changing the existing API might not be an option here.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkersController.java:190]: This is path that should never hit as the dart side always serializes markerId to the marker object. I'll add test to this in separate commit if possible.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GoogleMapController.java:930]: Added test cases to cover these methods

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/utils/cluster_manager_utils.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/utils/cluster_manager_utils.dart:8]: Added tests for cluster_manager_utils

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:187]: Please check how this _events method is used elsewhere in the file.
_events function just creates event matcher that matches the mapId, and creating stream with invalid mapId will just create stream that does not match any events ever.
The stream is created for the current mapId and I don't see case where the mapId could be invalid.
Testing this method would test mostly the dart Stream functionality.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (camsim99) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:166]: Nit: Maybe just point to the class/method this is in versus the link?

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/utils/cluster_manager_utils.dart
Comment (camsim99) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/utils/cluster_manager_utils.dart:15]: Nit: can this be private? Don't see it used anywhere else

**Review by camsim99 — APPROVED** [review]
Left some nits but all of my comments have been addressed! Looks good :)

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:92]: @reidbaker @stuartmorgan do you still need changes to this?

Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/ClusterManagersController.java:92]: I'm fine with this staying as is for now since a) it should never actually happen, and b) it will go away when we convert to Pigeon, which I plan to do this quarter.

**Review by camsim99 — APPROVED** [review]
Thanks for fixing my nits! @reidbaker for second review

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerController.java
Comment (reidbaker) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/MarkerController.java:24]: yes that comment works for me.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart
Comment (reidbaker) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:187]: Ok so what you are saying is that an invalid map id is a user error and is safe to call?

**Review by reidbaker — APPROVED** [review]

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart
Comment (jokerttu) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:187]: Yes it is. The app facing plugin implementation is also hiding the map id from the developer interface.
If anyone uses these plarform interface implementations directly, it is user error to listen wron mapId.
Note that this whole multiplexing behaviour will be removed when this plugin is transformed to use pigeon messaging.

**Pull request merged**
